### PR TITLE
Update middleware function to correctly honor trace headers

### LINKED ISSUE #117: Ensure Correct Xmidt Header Usage
In `middleware.go`, Candlelight uses the incorrect header format: `X-Midt-*`. This should be updated to use the correct header format: `X-Xmidt-*`

### PR DESCRIPTION
**Pull request by renaz6**
- With the newest Tr1d1um release, tracing headers can now be sent in WRP messages (which can be sent as Xmidt headers, JSON, or msgpack)
- This PR updates Candlelight with capability to get the trace headers (if they exist) using the new `wrpcontext` package. Candlelight then propagates the trace context with this info so that remote spans can be honored and not broken.
- Also updated the header names to have the correct prefix. Closes #117

### CODE REVIEW

File: middleware.go [hunk 1]
```diff
@@ -73,6 +89,60 @@ func EchoFirstTraceNodeInfo(propagator propagation.TextMapPropagator) func(http.
...
+	} else {
+
+		// Check for trace info in the request body
+		reader := r.Body
+		bodyData, err := io.ReadAll(reader)
+		if err != nil {
+			log.Printf("failed to read body: %s", err)
+		}
+
+		r.Body, r.GetBody = xhttp.NewRewindBytes(bodyData)
```
Comment (schmidtw) [inline middleware.go:110]: This is the only function that is needed from webpa-common ... but it's a bit tangled.  The `r.Body` is a `io.ReadCloser` so I'm not sure if the `bufio.Reader` is going to work here.  @johnabass any suggestions?

Comment (johnabass) [reply middleware.go]: We should defiinitely remove the rewind stuff here.  It's really just an optimization hack to allow reuse of the same buffer when redirects happen.  I would first try to make `bufio.Reader` work, in tandem with `io.NopCloser`.  Failing that, we should move this to `httpaux`.

File: middleware.go [hunk 2]
```diff
@@ -73,6 +89,66 @@ func EchoFirstTraceNodeInfo(propagator propagation.TextMapPropagator) func(http.
...
+			_, err := newBody.Seek(0, 0)
+			return io.NopCloser(newBody), err
+		}
+
+		var msg wrp.Message
+		if err := json.Unmarshal(bodyData, &msg); err != nil {
+			log.Printf("failed to unmarshal payload: %s", err)
+		}
+
+		if r.Header.Get("Content-Type") == "application/msgpack" {
```
Comment (johnabass) [inline middleware.go:123]: The knowledge of what content types are associated with which `wrp` formats more properly belongs in `wrp-go`.  We have some of that functionality already: https://pkg.go.dev/github.com/xmidt-org/wrp-go@v1.3.4/wrp#FormatFromContentType

What's missing is a way to fully extract a `wrp` message from a request, using the `context` as a "cache".  I've added the following issues to `wrp-go` to address this:

- https://github.com/xmidt-org/wrp-go/issues/114
- https://github.com/xmidt-org/wrp-go/issues/115

**Review by johnabass — CHANGES_REQUESTED** [review]

File: middleware.go (hunk 2 above)
Comment (renaz6) [reply middleware.go]: Updated in https://github.com/xmidt-org/wrp-go/pull/116/

File: middleware.go [hunk 3]
```diff
@@ -62,11 +67,31 @@ func (traceConfig *TraceConfig) TraceMiddleware(delegate http.Handler) http.Hand
 func EchoFirstTraceNodeInfo(propagator propagation.TextMapPropagator) func(http.Handler) http.Handler {
 	return func(delegate http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+			r, _ = wrphttp.DecodeRequest(r, nil)
```
Comment (denopink) [inline middleware.go:71]: assuming this error state doesn't matter

File: middleware.go [hunk 4]
```diff
@@ -62,11 +67,31 @@ func (traceConfig *TraceConfig) TraceMiddleware(delegate http.Handler) http.Hand
...
+			// Iterate through the trace headers (if any), format them, and add them to ctx
+			var tmp propagation.TextMapCarrier = propagation.MapCarrier{}
+			for _, f := range traceHeaders {
+				if f != "" {
+					parts := strings.Split(f, ":")
+					// Remove leading space if there's any
+					parts[1] = strings.Trim(parts[1], " ")
+					tmp.Set(parts[0], parts[1])
+				}
+				ctx = propagation.TraceContext{}.Extract(ctx, tmp)
```
Comment (denopink) [inline middleware.go:88]: Don't you want to do this step , `ctx = propagation.TraceContext{}.Extract(ctx, tmp) `, after going through all the trace headers `f`'s ?

**Review by denopink — APPROVED** [review]
left some questions, but overall lgtm 🍻

File: middleware.go (hunk 4 above)
Comment (renaz6) [reply middleware.go]: Great point. I can definitely just do this once. Updated in latest commit.

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+- Add ability to extract trace headers (traceparent and tracestate) from Xmidt headers, JSON, and msgpack and propagates the trace context accordingly [#115](https://github.com/xmidt-org/candlelight/pull/115)
```
Comment (renaz6) [inline CHANGELOG.md:8]: @schmidtw I know we don't need to update the Changelogs anymore, but is it still good practice?

Comment (schmidtw) [reply CHANGELOG.md:8]: It's probably better to update when the file is there.

**Review by schmidtw — APPROVED** [review]

**Pull request merged**